{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_attribute(attr):\n",
    "    '''get gene_id from attribute\n",
    "    '''\n",
    "    for item in attr.split(';'):\n",
    "        key, value = item.split('=')\n",
    "        if key == 'ID':\n",
    "            return value.replace('gene:', '')\n",
    "\n",
    "\n",
    "def get_gene_is_reverse(gff_path):\n",
    "    GFF_COLUMNS = ['seqid', 'source', 'type', 'start', 'end', \n",
    "                   'score', 'strand', 'phase', 'attribute']\n",
    "    gff_df = pd.read_csv(gff_path, sep='\\t', comment='#', names=GFF_COLUMNS)\n",
    "    # keep type = 'gene'\n",
    "    gff_df = gff_df[gff_df['type'].isin(['gene'])]\n",
    "    # get gene_id\n",
    "    gff_df['gene_id'] = gff_df['attribute'].map(parse_attribute)\n",
    "    gff_df['is_reverse'] = gff_df['strand'].map(lambda x: False if x=='+' else True)\n",
    "    \n",
    "    return gff_df.set_index('gene_id')['is_reverse'].T.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_polya_data(polya_data):\n",
    "    # polya_data = '/public/home/mowp/test/fast5_api/fhh.polyacaller.ann.tsv'\n",
    "    df = pd.read_csv(polya_data, sep='\\t', index_col=0)\n",
    "    df['read_is_reverse'] = df['read_type'].map(lambda x: True if x == 'polyT' else False)\n",
    "\n",
    "    # a dict like this: {'read_id': True or False}\n",
    "    read_is_reverse = df['read_is_reverse'].to_dict()\n",
    "    # a dict like this: {'read_id': 'gene_id'}\n",
    "    read_gene_id = df['gene_id'].to_dict()\n",
    "    # a dict like this: {'read_id': 100.0}\n",
    "    read_tail_length = df['tail_length'].to_dict()\n",
    "    return read_is_reverse, read_gene_id, read_tail_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "gff_path = '/public/home/mowp/db/Arabidopsis_thaliana/gff3/Arabidopsis_thaliana.TAIR10.46.gff3'\n",
    "gene_is_reverse_dict = get_gene_is_reverse(gff_path)\n",
    "\n",
    "polya_data = '/public/home/mowp/test/fast5_api/fhh.polyacaller.ann.tsv'\n",
    "read_is_reverse_dict, read_gene_id_dict, read_tail_length_dict = load_polya_data(polya_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "infile = '/public/home/mowp/test/nanopore_cdna/aligned_data/fhh.adjust.mm2.sorted.bam'\n",
    "outfile = '/public/home/mowp/test/nanopore_cdna/aligned_data/fhh.tagged.mm2.sorted.bam'\n",
    "with pysam.AlignmentFile(infile, 'rb') as inbam, pysam.AlignmentFile(outfile, 'wb', template=inbam) as outbam:\n",
    "    # add tags: pa for polya_length\n",
    "    #           gi for gene_id\n",
    "    for read in inbam.fetch():\n",
    "        # only output mapped reads\n",
    "        if not read.is_unmapped:\n",
    "            read_is_reverse = read_is_reverse_dict[read.query_name]\n",
    "            read_gene_id = read_gene_id_dict[read.query_name]\n",
    "            read_tail_length = read_tail_length_dict[read.query_name]\n",
    "            # 能被注释到gene_id的执行如下操作\n",
    "            try:\n",
    "                gene_is_reverse = gene_is_reverse_dict[read_gene_id]\n",
    "                read.set_tag('gi', read_gene_id)\n",
    "                if gene_is_reverse is read.is_reverse:\n",
    "                    read.set_tag('pa', read_tail_length)\n",
    "                else:\n",
    "                    continue\n",
    "            # 不能注释的直接输出，有可能是新的转录本\n",
    "            except KeyError:\n",
    "                read.set_tag('gi', 'None')\n",
    "                read.set_tag('pa', read_tail_length)\n",
    "            outbam.write(read)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "115.13635194237202"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read_tail_length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('NM', 69),\n",
       " ('ms', 1499),\n",
       " ('AS', 1331),\n",
       " ('nn', 0),\n",
       " ('ts', '+'),\n",
       " ('tp', 'P'),\n",
       " ('cm', 420),\n",
       " ('s1', 1495),\n",
       " ('s2', 0),\n",
       " ('de', 0.02329999953508377),\n",
       " ('cs',\n",
       "  ':23-ag:146+g:21+g:1*ca*ga:9+c:54*tc:2-cta:2-a:17~gt82ag:23-g:49*ag:53-g:54-a:6-t:88+c:3~gt209ag:30+aa:42+c:3*at:27*gt:16~gt100ag:133-ct:48-ta:183-tga:19~gt78ag:17*ga*ag:134~gt112ag:27-g:26-a:71-a:76*ga:1-a:5*ga:18+ag*tc:10-ttt:7-tt:62-aaagaagagaaaaaggggg:45+a:44-a:1*tg:34*ga:3'),\n",
       " ('rl', 84),\n",
       " ('pa', 115.1363525390625),\n",
       " ('ge', 'AT1G01010'),\n",
       " ('gi', nan)]"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "read.get_tags()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>read_type</th>\n",
       "      <th>tail_start</th>\n",
       "      <th>tail_end</th>\n",
       "      <th>samples_per_nt</th>\n",
       "      <th>tail_length</th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>read_is_reverse</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>read_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>84778c00-f131-4c7e-a741-a3d17a72ae95</th>\n",
       "      <td>polyT</td>\n",
       "      <td>1176</td>\n",
       "      <td>2572</td>\n",
       "      <td>9.293564</td>\n",
       "      <td>176.251014</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39f88611-6c7f-439d-a1bc-4baff95c9eb8</th>\n",
       "      <td>polyA</td>\n",
       "      <td>6094</td>\n",
       "      <td>7480</td>\n",
       "      <td>9.681818</td>\n",
       "      <td>144.394366</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4d6c098a-e098-423a-a315-6c45a3b51b28</th>\n",
       "      <td>polyA</td>\n",
       "      <td>2334</td>\n",
       "      <td>2706</td>\n",
       "      <td>9.648069</td>\n",
       "      <td>88.722420</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>558d297a-c033-4663-9696-525c96f2946e</th>\n",
       "      <td>polyT</td>\n",
       "      <td>2288</td>\n",
       "      <td>2924</td>\n",
       "      <td>10.660920</td>\n",
       "      <td>61.533154</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>314c0d3a-3bec-4989-bbb0-737841bd535c</th>\n",
       "      <td>polyT</td>\n",
       "      <td>830</td>\n",
       "      <td>1052</td>\n",
       "      <td>9.890995</td>\n",
       "      <td>24.062290</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f56fe844-910b-4c46-8520-46afbe8739ea</th>\n",
       "      <td>polyA</td>\n",
       "      <td>13616</td>\n",
       "      <td>16026</td>\n",
       "      <td>9.446951</td>\n",
       "      <td>279.455253</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>f91087bc-6a2c-4212-8cc3-ccb285d55583</th>\n",
       "      <td>polyT</td>\n",
       "      <td>986</td>\n",
       "      <td>1116</td>\n",
       "      <td>10.048059</td>\n",
       "      <td>14.132082</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9b91b8c2-ec34-496d-b96e-027fca06edd6</th>\n",
       "      <td>polyT</td>\n",
       "      <td>1512</td>\n",
       "      <td>1512</td>\n",
       "      <td>10.549751</td>\n",
       "      <td>2.274935</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0b71cd88-37c7-4531-aeb2-d1632c22999b</th>\n",
       "      <td>polyT</td>\n",
       "      <td>1148</td>\n",
       "      <td>2352</td>\n",
       "      <td>8.985577</td>\n",
       "      <td>136.218299</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>e4740ef4-1bbb-487e-b409-58c5e85dfa1f</th>\n",
       "      <td>polyA</td>\n",
       "      <td>5584</td>\n",
       "      <td>6610</td>\n",
       "      <td>9.314410</td>\n",
       "      <td>111.440225</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>69646 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     read_type  tail_start  tail_end  \\\n",
       "read_id                                                                \n",
       "84778c00-f131-4c7e-a741-a3d17a72ae95     polyT        1176      2572   \n",
       "39f88611-6c7f-439d-a1bc-4baff95c9eb8     polyA        6094      7480   \n",
       "4d6c098a-e098-423a-a315-6c45a3b51b28     polyA        2334      2706   \n",
       "558d297a-c033-4663-9696-525c96f2946e     polyT        2288      2924   \n",
       "314c0d3a-3bec-4989-bbb0-737841bd535c     polyT         830      1052   \n",
       "...                                        ...         ...       ...   \n",
       "f56fe844-910b-4c46-8520-46afbe8739ea     polyA       13616     16026   \n",
       "f91087bc-6a2c-4212-8cc3-ccb285d55583     polyT         986      1116   \n",
       "9b91b8c2-ec34-496d-b96e-027fca06edd6     polyT        1512      1512   \n",
       "0b71cd88-37c7-4531-aeb2-d1632c22999b     polyT        1148      2352   \n",
       "e4740ef4-1bbb-487e-b409-58c5e85dfa1f     polyA        5584      6610   \n",
       "\n",
       "                                      samples_per_nt  tail_length  \\\n",
       "read_id                                                             \n",
       "84778c00-f131-4c7e-a741-a3d17a72ae95        9.293564   176.251014   \n",
       "39f88611-6c7f-439d-a1bc-4baff95c9eb8        9.681818   144.394366   \n",
       "4d6c098a-e098-423a-a315-6c45a3b51b28        9.648069    88.722420   \n",
       "558d297a-c033-4663-9696-525c96f2946e       10.660920    61.533154   \n",
       "314c0d3a-3bec-4989-bbb0-737841bd535c        9.890995    24.062290   \n",
       "...                                              ...          ...   \n",
       "f56fe844-910b-4c46-8520-46afbe8739ea        9.446951   279.455253   \n",
       "f91087bc-6a2c-4212-8cc3-ccb285d55583       10.048059    14.132082   \n",
       "9b91b8c2-ec34-496d-b96e-027fca06edd6       10.549751     2.274935   \n",
       "0b71cd88-37c7-4531-aeb2-d1632c22999b        8.985577   136.218299   \n",
       "e4740ef4-1bbb-487e-b409-58c5e85dfa1f        9.314410   111.440225   \n",
       "\n",
       "                                     transcript_id gene_id  read_is_reverse  \n",
       "read_id                                                                      \n",
       "84778c00-f131-4c7e-a741-a3d17a72ae95           NaN     NaN             True  \n",
       "39f88611-6c7f-439d-a1bc-4baff95c9eb8           NaN     NaN            False  \n",
       "4d6c098a-e098-423a-a315-6c45a3b51b28           NaN     NaN            False  \n",
       "558d297a-c033-4663-9696-525c96f2946e           NaN     NaN             True  \n",
       "314c0d3a-3bec-4989-bbb0-737841bd535c           NaN     NaN             True  \n",
       "...                                            ...     ...              ...  \n",
       "f56fe844-910b-4c46-8520-46afbe8739ea           NaN     NaN            False  \n",
       "f91087bc-6a2c-4212-8cc3-ccb285d55583           NaN     NaN             True  \n",
       "9b91b8c2-ec34-496d-b96e-027fca06edd6           NaN     NaN             True  \n",
       "0b71cd88-37c7-4531-aeb2-d1632c22999b           NaN     NaN             True  \n",
       "e4740ef4-1bbb-487e-b409-58c5e85dfa1f           NaN     NaN            False  \n",
       "\n",
       "[69646 rows x 8 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'84778c00-f131-4c7e-a741-a3d17a72ae95'"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
